### PR TITLE
[debugger] Invoke method with fixed size array as an attribute of this

### LINKED ISSUE #15556: [sdb] Invocation issue with fixed buffers
Unzip the following repro:

[reprosdb5.zip](https://github.com/mono/mono/files/3357049/reprosdb5.zip)

`./build.sh && mono debugger.exe`

```
lemacbook:reprosdb5 jbevain$ mono debugger.exe 
Launching debuggee.exe
VMStart
Resuming debuggee.exe
TypeLoad
TypeLoaded: Program
Method: Void Program:Main ()
(1, 2, 3, 4)
Bye
Breakpoint
(1, -9984, 32722, 0)
VMDeath
Bye felicia
```

The line `(1, -9984, 32722, 0)` will vary and shows the issue.
It should read `(1, 2, 3, 4)`, however, I have the feeling that the `decode_value` / `decode_vtype` marshaling mechanism doesn't handle fixed buffers or just packed structs, and we end up reading garbage.

### PR DESCRIPTION
**Pull request by thaystg**
Implement debugger invoke when the this has an attribute that is a fixed size array.

Fixes #15556

### CODE REVIEW

File: mcs/class/Mono.Debugger.Soft/Test/dtest.cs [hunk 1]
```diff
@@ -71,8 +71,8 @@ void StackTraceDump (Event e)
 #if !MONODROID_TOOLS
 	Diag.ProcessStartInfo CreateStartInfo (string app, string method = null, string runtimeParameters = null) {
 		var pi = new Diag.ProcessStartInfo ();
-		pi.RedirectStandardOutput = true;
-		pi.RedirectStandardError = true;
+		//pi.RedirectStandardOutput = true;
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: Unrelated change.

File: mcs/class/Mono.Debugger.Soft/Test/dtest.cs [hunk 2]
```diff
@@ -5044,6 +5044,48 @@ public void Bug59649 ()
...
+		var req = create_step (e);
+		req.Enable ();
+		step_once ();
+		step_over ();
+		step_over ();
+		var bp = step_over ();
+		var thread = bp.Thread;
+		var frame = thread.GetFrames()[0];
+
+		var local =  frame.GetVisibleVariableByName("n");
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: Where does this test fixed size arrays ?

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 3]
```diff
@@ -853,36 +854,228 @@ class PacketReader {
 				return res;
 			}
 
+			public int ReadFixedSize () {
+				if (connection.Version.AtLeast (2, 53)) {
+					bool isFixedSize = ReadByte () == 1;
+					int lenFixedSize = 1;
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: This will increase the size of every value returned by the debugger by 1 byte, is this really needed ?

File: mcs/class/Mono.Debugger.Soft/Test/dtest.cs (hunk 2 above)
Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Test/dtest.cs]: Here:
public unsafe struct NodeTestFixedArray
{
	private fixed short buffer[4];
	private fixed char buffer2[4];

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs (hunk 3 above)
Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: The other solution is to create a new datatype like, MONO_TYPE_FIXED_SIZE, and then we read the datatype of fixed sized array later, if you think this is better, just let me know...and I change...

Comment (vargaz) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: I think that would be better, this patch is like 1000 lines just to support this rarely used feature.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 4]
```diff
@@ -853,10 +855,233 @@ class PacketReader {
...
+					return new ValueImpl { Type = etype, Value = ReadDoubleValue () };
+				}
+				throw new NotImplementedException ("Unable to handle type " + etype);
+			}
 			public ValueImpl ReadValue () {
 				ElementType etype = (ElementType)ReadByte ();
-
 				switch (etype) {
+				case (ElementType)ValueTypeId.VALUE_TYPE_ID_FIXED_ARRAY:
+					return ReadValueFixedSize ();
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: This could be simplified now since we know its a fixed array, no need for the Read<something>Value methods.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 5]
```diff
@@ -880,7 +1105,7 @@ class PacketReader {
 				case ElementType.U8:
 					return new ValueImpl { Type = etype, Value = (ulong)ReadLong () };
 				case ElementType.R4:
-					return new ValueImpl { Type = etype, Value = ReadFloat () };
+					return new ValueImpl { Type = etype, Value = ReadFloat ()  };
 				case ElementType.R8:
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Extra whitespace.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 6]
```diff
@@ -1042,74 +1267,146 @@ class PacketWriter {
...
-				else
+				if (v.Value != null) {
+					t = TypeCodeToElementType (Type.GetTypeCode (v.Value.GetType ()), v.Value.GetType ());
+				}
+				else {
 					t = v.Type;
+				}
+				if (v.FixedSize > 1 && t != ElementType.ValueType)
+				{
+					WriteFixedSizeValue (v);
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Same here.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 7]
```diff
@@ -853,10 +855,94 @@ class PacketReader {
...
+						return new ValueImpl { Type = etype, Value = val };
+					}
+				}
+				throw new NotImplementedException ("Unable to handle type " + etype);
+			}
 
+			public ValueImpl ReadValue () {
+				ElementType etype = (ElementType)ReadByte ();
 				switch (etype) {
+				case (ElementType)ValueTypeId.VALUE_TYPE_ID_FIXED_ARRAY:
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: I would put this case at the end along with the other ValueTypeId cases.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 8]
```diff
@@ -1042,74 +1128,138 @@ class PacketWriter {
 				return this;
 			}
 
+			public PacketWriter WriteFixedSizeValue (ValueImpl v) {
+				ElementType t;
+
+				if (v.Value != null) {
+					t = TypeCodeToElementType (Type.GetTypeCode (v.Value.GetType ()), v.Value.GetType ());
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Extra braces.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 9]
```diff
@@ -1042,74 +1128,138 @@ class PacketWriter {
...
+				}
+				else {
+					t = v.Type;
+				}
+				WriteByte ((byte) ValueTypeId.VALUE_TYPE_ID_FIXED_ARRAY);
+				WriteByte ((byte)t);
+				WriteInt (v.FixedSize);
+				for (int j = 0 ; j < v.FixedSize; j++)
+				{
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Should use K&R indentation.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 10]
```diff
@@ -1042,74 +1128,138 @@ class PacketWriter {
...
+			}
+
 			public PacketWriter WriteValue (ValueImpl v) {
 				ElementType t;
 
-				if (v.Value != null)
-					t = TypeCodeToElementType (Type.GetTypeCode (v.Value.GetType ()));
-				else
+				if (v.Value != null) {
+					t = TypeCodeToElementType (Type.GetTypeCode (v.Value.GetType ()), v.Value.GetType ());
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Extra braces.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 11]
```diff
@@ -1042,74 +1128,138 @@ class PacketWriter {
...
+				}
+				else {
 					t = v.Type;
+				}
+				if (v.FixedSize > 1 && t != ElementType.ValueType)
+				{
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Should use K&R indentation.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 12]
```diff
@@ -1042,74 +1128,138 @@ class PacketWriter {
...
+					WriteFixedSizeValue (v);
+					return this;
+				}
 				WriteByte ((byte)t);
 				switch (t) {
-				case ElementType.Boolean:
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: This section got re-indented for some reason.

**Comment by vargaz** [comment]
Looks ok otherwise.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 13]
```diff
@@ -215,6 +216,7 @@ class ValueImpl {
 		public bool IsEnum; // For ElementType.ValueType
 		public long Id; /* For VALUE_TYPE_ID_TYPE */
 		public int Index; /* For VALUE_TYPE_PARENT_VTYPE */
+		public int FixedSize = 1; 
 	}
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Why does this need to be inited to 1 ?

File: mono/mini/debugger-agent.c [hunk 14]
```diff
@@ -5394,6 +5490,7 @@ obj_is_of_type (MonoObject *obj, MonoType *t)
 	return TRUE;
 }
 
+
 static ErrorCode
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: Extra whitespace.

File: mono/mini/debugger-agent.c [hunk 15]
```diff
@@ -5206,6 +5296,7 @@ buffer_add_value_full (Buffer *buf, MonoType *t, void *addr, MonoDomain *domain,
 		addr = *(void**)addr;
 	}
 
+
 	if (as_vtype) {
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: Extra whitespace.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs (hunk 13 above)
Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: We don't need anymore, we need this before when I used the same method for both cases.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs (hunk 7 above)
Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: okay!

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs (hunk 8 above)
Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: okay.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs (hunk 11 above)
Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: okay

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 16]
```diff
@@ -916,10 +917,95 @@ class PacketReader {
...
 					return new ValueImpl () { Type = etype, Index = ReadInt () };
+				case (ElementType)ValueTypeId.VALUE_TYPE_ID_FIXED_ARRAY:
+					return ReadValueFixedSize ();
 				default:
 					throw new NotImplementedException ("Unable to handle type " + etype);
 				}
 			}
+			
+			public ValueImpl ReadValueFixedSize () {
+				var lenFixedSize = 1;
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: This doesn't have to be public.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 17]
```diff
@@ -916,10 +917,95 @@ class PacketReader {
...
+				ElementType etype = (ElementType)ReadByte ();
+				if (connection.Version.AtLeast (2, 53)) {
+					lenFixedSize = ReadInt ();
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Older protocol versions can't send the type code which would cause this method to be reached.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 18]
```diff
@@ -916,10 +917,95 @@ class PacketReader {
...
+						var val = new sbyte[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+							val[i] = (sbyte)ReadInt ();
+						return new ValueImpl { Type = etype, Value = val };
+					}
+					case ElementType.U1: {
+						var val = new byte[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+								val[i] = (byte)ReadInt ();
+						return new ValueImpl { Type = etype, Value = val };
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Bad indentation.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 19]
```diff
@@ -916,10 +917,95 @@ class PacketReader {
...
+					}
+					case ElementType.Boolean: {
+						var val = new bool[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+								val[i] = (bool) (ReadInt () != 0);
+						return new ValueImpl { Type = etype, Value = val };
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Ditto.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 20]
```diff
@@ -916,10 +917,95 @@ class PacketReader {
...
+					}
+					case ElementType.I2: {
+						var val = new short[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+								val[i] = (short)ReadInt ();
+						return new ValueImpl { Type = etype, Value = val };
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Ditto.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 21]
```diff
@@ -916,10 +917,95 @@ class PacketReader {
...
+						var val = new int[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+							val[i] = ReadInt ();
+						return new ValueImpl { Type = etype, Value = val };
+					}
+					case ElementType.U4: {
+						var val = new uint[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+								val[i] = (uint)ReadInt ();
+						return new ValueImpl { Type = etype, Value = val };
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Ditto.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 22]
```diff
@@ -916,10 +917,95 @@ class PacketReader {
...
+						var val = new float[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+							val[i] = ReadFloat ();
+						return new ValueImpl { Type = etype, Value = val };
+					}
+					case ElementType.R8: {
+						var val = new double[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+								val[i] = ReadDouble ();
+						return new ValueImpl { Type = etype, Value = val };
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Ditto.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/FieldInfoMirror.cs [hunk 23]
```diff
@@ -162,6 +162,21 @@ public Boolean IsNotSerialized
 			}
 		}
 
+		public int IsFixedSize
+		{
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/FieldInfoMirror.cs]: If its not a boolean, it shouldn't be named IsSomething

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 24]
```diff
@@ -1113,6 +1203,61 @@ class PacketWriter {
 				return this;
 			}
 
+			public PacketWriter WriteFixedSizeValue (ValueImpl v) {
+				ElementType t;
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: This doesn't need to be public.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/FieldInfoMirror.cs (hunk 23 above)
Comment (vargaz) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/FieldInfoMirror.cs]: Is this new public property needed ?

Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/FieldInfoMirror.cs]: Yes, it's not a boolean, it's the size, but I need to access it to decide on WriteValue if it's a fixed size array.

Comment (vargaz) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/FieldInfoMirror.cs]: The property doesn't need to be public. Also, the return value needs to be cached, since this call might make multiple roundtrips to the debuggee. Also, it should be possible to just return false if the type of the field is not a valuetype.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs [hunk 25]
```diff
@@ -916,11 +917,94 @@ class PacketReader {
...
+					case ElementType.U1: {
+						var val = new byte[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+							val[i] = (byte)ReadInt ();
+						return new ValueImpl { Type = etype, Value = val };
+					}
+					case ElementType.Boolean: {
+						var val = new bool[lenFixedSize];
+						for (int i = 0; i < lenFixedSize; i++)
+							val[i] = (bool) (ReadInt () != 0);
```
Comment (jaykrell) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Why cast? Isn't `x != 0` already `bool`?

File: mono/mini/debugger-agent.c [hunk 26]
```diff
@@ -3188,6 +3196,44 @@ process_frame (StackFrameInfo *info, MonoContext *ctx, gpointer user_data)
...
+{
+	ERROR_DECL (error);
+	if (!CHECK_PROTOCOL_VERSION (2, 53)) {
+		return 1;
+	}
+	MonoCustomAttrInfo *cinfo;
+	MonoCustomAttrEntry *attr;
+	int aindex;
+	gint32 ret = 1;
+	cinfo = mono_custom_attrs_from_field_checked (f->parent, f, error);
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: This should be avoided if the field type is not a valuetype.

File: mono/mini/debugger-agent.c [hunk 27]
```diff
@@ -546,6 +548,12 @@ typedef struct ReplyPacket {
 #define CHECK_PROTOCOL_VERSION(major,minor) \
 	(protocol_version_set && (major_version > (major) || (major_version == (major) && minor_version >= (minor))))
 
+#define BUFFER_ADD_PRIMITIVE_VALUE() \
+	buffer_add_byte (buf, t->type); \
+	if (CHECK_PROTOCOL_VERSION (2, 53)) { \
+		buffer_add_int (buf, len_fixed_array ); \
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: This should be checked in the caller once, not in a macro.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs (hunk 25 above)
Comment (thaystg) [reply mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs]: Yes!

File: mono/mini/debugger-agent.c [hunk 28]
```diff
@@ -3188,6 +3190,44 @@ process_frame (StackFrameInfo *info, MonoContext *ctx, gpointer user_data)
...
+	gint32 ret = 1;
+	cinfo = mono_custom_attrs_from_field_checked (f->parent, f, error);
+	if (!is_ok (error))
+		return ret;
+	attr = NULL;
+	if (cinfo) {
+		for (aindex = 0; aindex < cinfo->num_attrs; ++aindex) {
+			MonoClass *ctor_class = cinfo->attrs [aindex].ctor->klass;
+			if (mono_class_has_parent (ctor_class,  mono_class_load_from_name (mono_defaults.corlib, "System.Runtime.CompilerServices", "FixedBufferAttribute"))) {
+				attr = &cinfo->attrs [aindex];
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: The class needs to be cached as well, see the GENERATE_GET_CLASS_WITH_CACHE macro.

**Comment by thaystg** [comment]
@monojenkins build failed

File: mono/mini/debugger-agent.c [hunk 29]
```diff
@@ -720,6 +722,8 @@ static void process_breakpoint_events (void *_evts, MonoMethod *method, MonoCont
 static int ss_create_init_args (SingleStepReq *ss_req, SingleStepArgs *args);
 static void ss_args_destroy (SingleStepArgs *ss_args);
 
+static GENERATE_GET_CLASS_WITH_CACHE (fixed_buffer, "System.Runtime.CompilerServices", "FixedBufferAttribute")
```
Comment (lambdageek) [inline mono/mini/debugger-agent.c]: maybe `GENERATE_TRY_GET_CLASS_WITH_CACHE` is better in case the `FixedBufferAttribute` class was removed by the mono linker?

File: mono/mini/debugger-agent.c [hunk 30]
```diff
@@ -3188,6 +3192,44 @@ process_frame (StackFrameInfo *info, MonoContext *ctx, gpointer user_data)
...
+	int aindex;
+	gint32 ret = 1;
+	cinfo = mono_custom_attrs_from_field_checked (f->parent, f, error);
+	if (!is_ok (error))
+		return ret;
+	attr = NULL;
+	if (cinfo) {
+		for (aindex = 0; aindex < cinfo->num_attrs; ++aindex) {
+			MonoClass *ctor_class = cinfo->attrs [aindex].ctor->klass;
+			if (mono_class_has_parent (ctor_class,  mono_class_get_fixed_buffer_class ())) {
```
Comment (lambdageek) [inline mono/mini/debugger-agent.c]: if the `FixedBufferAttribute` class is linked away, `mono_class_get_fixed_buffer_class` will return NULL and then the call to `mono_class_has_parent` will crash - it doesn't do a null check.

File: mono/mini/debugger-agent.c (hunk 29 above)
Comment (thaystg) [reply mono/mini/debugger-agent.c]: Done!

File: mono/mini/debugger-agent.c (hunk 30 above)
Comment (thaystg) [reply mono/mini/debugger-agent.c]: Done

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins commit apidiff

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

File: mono/mini/debugger-agent.c [hunk 31]
```diff
@@ -5263,7 +5347,12 @@ buffer_add_value_full (Buffer *buf, MonoType *t, void *addr, MonoDomain *domain,
 			break;
 		}
 	}
-
+	
+	if (len_fixed_array > 1 && t->type != MONO_TYPE_VALUETYPE && CHECK_PROTOCOL_VERSION (2, 53))
+	{
```
Comment (jbevain) [inline mono/mini/debugger-agent.c]: Nit: `{` should be on the `if` line.

File: mono/mini/debugger-agent.c [hunk 32]
```diff
@@ -5263,7 +5347,12 @@ buffer_add_value_full (Buffer *buf, MonoType *t, void *addr, MonoDomain *domain,
...
+		buffer_add_fixed_array(buf, t, addr, domain, as_vtype, parent_vtypes, len_fixed_array);
```
Comment (jbevain) [inline mono/mini/debugger-agent.c]: Nit: space before the parenthese.

File: mono/mini/debugger-agent.c [hunk 33]
```diff
@@ -3196,6 +3200,45 @@ process_frame (StackFrameInfo *info, MonoContext *ctx, gpointer user_data)
 	return FALSE;
 }
 
+static gint32 isFixedSizeArray (MonoClassField *f)
```
Comment (jbevain) [inline mono/mini/debugger-agent.c]: Nit: `is_fixed_size_array`
